pep-0612.rst

```diff
@@ -0,0 +1,258 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A ``ParameterSpecification`` captures both positional and keyword accessible parameters, but there unfortunately is no object in the runtime that captures both of these together. Instead, we are forced to separate them into ``*args`` and ``**kwargs``\ , respectively. This means we need to be able to split apart a single ``ParameterSpecification`` into these two components, and then bring them back together into a call.  To do this, we introduce ``TParams.args`` to represent the tuple of positional arguments in a given call and ``TParams.kwargs`` to represent the corresponding ``Mapping`` of keywords to values. These operators can only be used together, as the annotated types for ``*args`` and ``**kwargs`` .
+
+.. code-block::
+
+   class G(Generic[TParams]):
+       def foo(*args: TParams.args, **kwargs: TParams.kwargs) -> int: ...       # Accepted
+       def bar(*args: TParams.kwargs, **kwargs: TParams.args) -> int: ...       # Rejected
+       def baz(*args: TParams.args) -> int: ...                                 # Rejected
+       stored_arguments: TParams.args                                           # Rejected
+       def bap(x: TParams.args) -> int: ...                                     # Rejected
+       def bop(*args: List[TParams.args], **kwargs: TParams.kwargs) -> int: ... # Rejected
+
+Because the default kind of parameter in Python (\ ``(x: int)``\ ) may be addressed both positionally and through its name, two valid invocations of a ``(*args: TParams.args, **kwargs: TParams.kwargs)`` function may give different partitions of the same set of parameters. Therefore we need to make sure that these special types are only brought into the world together, and are used together, so that our usage is valid for all possible partitions.
+
+With those requirements met, we can now take advantage of the unique properties afforded to us by this set up: 
+
+
+* Inside the function, ``args`` has the type ``TParams.args``\ , not ``Tuple[TParams.args, ...]`` as would be with a normal annotation (and likewise with the ``**kwargs``\ )
+* A function of type ``Callable[TParams, TReturn]`` can be called with ``(*args, **kwargs)`` if and only if ``args`` has the type ``TParams.args`` and ``kwargs`` has the type ``TParams.kwargs``\ , and that those types both originated from the same function declaration.
+* A function declared as 
+  ``def inner(*args: TParams.args, **kwargs: TParams.kwargs) -> X``
+  has type ``Callable[TParams, X]``.
+
+With these three properties, we now have the ability to fully type check parameter preserving decorators.
+
+One additional form that we want to support is functions that pass only a subset of their arguments on to another function. To avoid shadowing a named or keyword only argument in the ``ParameterSpecification`` we require that the additional arguments be anonymous arguments that precede the ``*args`` and ``*kwargs``
```

Comment:
What do you mean by `anonymous` arguments? Are those arguments prefixed by double underscore (`__`) as in the provided example. Is that rule enforced or suggested?

Reply:
This PR has been merged. For discussions about the substance of the proposal, please post to typing-sig.

(If you were half way through writing this when the merge went through, then my apologies; you'll want to repost to typing-sig to get responses.)